### PR TITLE
tool: correctly update initrd secrets

### LINKED ISSUE #103: New `boot.initrd.secrets` for existing initrd is not written to the ESP
Currently, the secret is always added to the initrd in the tempdir but not actually copied over to the ESP because AFAIK the secret does not change the input hash (and thus path) of the initrd in the nix store and in the ESP.

### PR DESCRIPTION
**Pull request by nikstur**
This PR is part of a larger series which is supposed to be merged in order but is broken up into multiple pieces to make review easier. It is supposed to be merged in this order:

#108 -> #107 -> #110 -> (#109)

To correctly overwrite existing initrd with newer secrets (from newer
generations), the links need to be sorted from oldest generation to
newest.

Closes #103 
Depends on #110

### CODE REVIEW

**Comment by blitz** [comment]
@nikstur I don't think this closes #104. :) Can you fix the PR description?

**Comment by nikstur** [comment]
Thx, good catch ;)

File: nix/tests/lanzaboote.nix [hunk 1]
```diff
@@ -145,6 +145,53 @@ in
...
+          };
+        };
+      };
+      testScript = ''
+        machine.start()
+        machine.wait_for_unit("multi-user.target")
+
+        # Assert that only two boot files exists (a single kernel and a single
+        # initrd). If there are two initrds, the test would not be able to test
+        # updating the secret of an already existing inird.
```
Comment (RaitoBezarius) [inline nix/tests/lanzaboote.nix:187]: initrd

File: nix/tests/lanzaboote.nix [hunk 2]
```diff
@@ -145,6 +145,53 @@ in
...
+          "/test" = originalSecret;
+        };
+        boot.initrd.postMountCommands = ''
+          cp /test /mnt-root/secret-from-initramfs
+        '';
+
+        specialisation.variant.configuration = {
+          boot.initrd.secrets = {
+            # For the test to evaluate, this needs to be forced for some reason.
+            "/test" = lib.mkForce newSecret;
```
Comment (RaitoBezarius) [inline nix/tests/lanzaboote.nix:177]: This needs to be forced because it's already in the parent specialisation.

Comment (RaitoBezarius) [reply nix/tests/lanzaboote.nix:177]: If you made `lib.mkDefault` in the parent, there would be no need for force I think.

**Review by RaitoBezarius — APPROVED** [review]
LGTM

**Comment by nikstur** [comment]
Thanks for the reviews @blitz and @RaitoBezarius. I'm happy that we got these reviewed and merged quickly!

**Comment by RaitoBezarius** [comment]
> Thanks for the reviews @blitz and @RaitoBezarius. I'm happy that we got these reviewed and merged quickly!

Thank you for your work and patience :)

**Pull request merged**